core/modules/css/KvInterfaceImplMySql.cc

```diff
@@ -0,0 +1,503 @@
+        throw CssError((boost::format("_getChildrenFullPath - failed to extract children error:%1% from query:%2%") % errObj.errMsg() % query).str());
+    }
+
+    return strVec;
+}
+
+
+
+void
+KvInterfaceImplMySql::deleteKey(std::string const& key) {
+    KvTransaction transaction(_conn);
+    _delete(key);
+    transaction.commit();
+}
+
+
+void
+KvInterfaceImplMySql::_delete(std::string const& key) {
+    // A transaction should be in process here.
+
+    // recursively delete child keys first
+    std::vector<string> childKeys = _getChildrenFullPath(key);
+
+    // trim off the parent key, leaving only the last item in the path.
+    for (std::vector<std::string>::iterator strItr = childKeys.begin(); strItr != childKeys.end(); ++strItr) {
+        _delete(*strItr);
+    }
+
+    boost::format fmQuery("DELETE FROM kvData WHERE kvKey='%1%'");
+    fmQuery % key;
+    std::string query = fmQuery.str();
+
+    // run query
+    sql::SqlErrorObject errObj;
+    sql::SqlResults resultsObj;
+    LOGF(_logger, LOG_LVL_DEBUG, "Executing query: %1%" % query);
+    if (not _conn.runQuery(query, resultsObj, errObj)) {
+        LOGF(_logger, LOG_LVL_ERROR, "SQL DELETE FROM failed: %1%" % query);
+        throw CssError(errObj);
+    }
+
+    // limit the row count to 1 - there should not be multiple rows for any key.
+    auto affectedRows = resultsObj.getAffectedRows();
+    if (affectedRows < 1) {
+        LOGF(_logger, LOG_LVL_ERROR, "SQL DELETE FROM failed (no such key): %1%" % query);
+        throw NoSuchKey(errObj);
+    } else if (affectedRows > 1) {
+        LOGF(_logger, LOG_LVL_ERROR, "SQL DELETE FROM failed (too many (%1%) rows deleted): %2%" % affectedRows % query);
+        throw CssError("deleteKey - unexpectedly deleted more than 1 row.");
+    }
+
+}
+
+
+std::string
+KvInterfaceImplMySql::_get(std::string const& key, std::string const& defaultValue, bool throwIfKeyNotFound) {
+    // Note regarding starting a transaction here:
+    // Most private methods (in fact all of them) require that a transaction already be in progress. On the other hand:
+    // this method is protected, and is called by KvInterface with no transaction in progress so here one must be started.
+    KvTransaction transaction(_conn);
+
+    std::string val;
+    sql::SqlErrorObject errObj;
+
+    boost::format fmQuery("SELECT kvVal FROM kvData WHERE kvKey='%1%'");
+    fmQuery % key;
+    std::string query = fmQuery.str();
+
+    sql::SqlResults results;
+    if (not _conn.runQuery(query, results, errObj)) {
+        LOGF(_logger, LOG_LVL_ERROR, "_get - query failed: %1%" % query);
+        if (throwIfKeyNotFound) {
+            throw CssError(errObj);
+        } else {
+            val = defaultValue;
+        }
+    } else {
+        errObj.reset();
+        if (not results.extractFirstValue(val, errObj)) {
+            if (throwIfKeyNotFound) {
```

Comment:
it's true. however very little information is provided by the sql module about what error occurred, so this does what it can with the info it has. I'll at least add a log line that prints the error message. To handle it programmatically we could set an error number to the errObj, inside SqlResults::extractFirstValue. I don't see error codes defined anywhere inside the sql module, should we add a namespace for that?